# Orobanchaceae_comparative_mitome
Scripts used in the comparative genome study of mitochondrial genome evolution in the parasitic Orobanchaceae

Citation: Recombination and retroprocessing in broomrapes reveal a universal roadmap for mitochondrial evolution in heterotrophic plants. L., Cai, J.C., Havird, R.K., Jansen. biorxiv. doi: https://doi.org/10.1101/2025.02.14.637881


Reach out to Liming Cai (daybreak.chua@gmail.com) for any questions

The data supporting the findings of this study, including genome assemblies, annotations, gene alignments, and phylogenies for the study of RNA editing, molecular evolution, and gene transfers are available in the Figshare Digital Repository (###).


Description of the subdirectories
----------------------
The content of each subdirectory is listed as follows. For detailed descriptions and scripts, please see the README file within each directory.

## 1_plastid_assembly_structure
  
  Plastid genome assembly

  Plastid IR region annotation
  
  Plastid reads removal for mitome assembly
## 2_mitome_assembly_structure

  Mitochondria genome assembly

  Annotation of repeat, synteny blocks, and mtpt

## 3_phylogeny

  Inference of mitochondrion- and plastid-based phylogeny

## 4_RNA_editing

  Deepredmt, mapping of loss of RNA editing sites, ancestral sequence reconstruction

## 5_mtpt

  Characterization of mitochondrial plastid DNA (MTPT)
  
## 6_HGT

  Mitochondrial HGT characterization

## 7_molecular_evolution

  Hyphy analyses, dN/dS calculation, organelle copy number

## 8_PGLS_PhylANOVA_COEVOL

  Comparative analysis and hypothesis testing using PGLS, phylogenetic ANOVA, and COEVOL 
  Comparative hypothesis testing using phylogenetic generalized least squares (PGLS), phylogenetic ANOVA (phylANOVA), and COEVOL.

## MISC

  Unused MISC analyses

